### DISCUSSION TITLE
Objective for individual community strains

### DISCUSSION THREAD

**Discussion by jmol0917** [micom-dev/micom#21] [Q&A] [answered]
Is there any way to modify the objective of individual members within a community? e.g. change it from biomass generation to specific metabolite generation?

I understand that community-level objectives can be selected by using ```com.objective = 'reaction_ID'``` but is there any option to do this for individual strains that are part of a community.

For example, if I have a community composed of Bacteroides caccae ATCC 43185 and Achromobacter xylosoxidans A8. Is there any way in which I can change the objective of one of these while the other's objective is still to generate biomass? Would modifying the objective of one of these strains using the ```cobrapy``` method ```model.objective``` work? or is there a way in which the objective can be specified in the dataframe that is passed when building the community: ```com = Community(taxonomy)```?

Thanks

**Comment by cdiener** [comment, marked as answer]
It's possible, however, I would not recommend using cooperative tradeoff in that case. ctFBA is only valid for growth rates. If you want to optimize growth and the production of a particular metabolite at the same time the recommended strategy would be to first run ctFBA and then maximize the target reaction while maintaining the obtained growth rates.

For instance:

```
from micom.data import test_taxonomy
from micom.util import _apply_min_growth
from micom import Community

tax = test_taxonomy()
com = Community(tax)
sol = com.cooperative_tradeoff()
with com:
    _apply_min_growth(com, sol.members.growth_rate.drop("medium"))
    com.objective = com.reactions.EX_ac_e__Escherichia_coli_1
    ac_sol = com.optimize()
print(ac_sol.objective_value)
```

If you just want to optimize the changed objective without any constraint on the growth rates you can do so the same way as in cobrapy and o things like: `com.objective = com.reactions.EX_ac_e__Escherichia_coli_1 + com.reactions.EX_ac_e__Escherichia_coli_3` or whatever you like. Note the added taxon name in the reaction ID.

Best

**Reply by jmol0917** [reply]
Amazing! Thanks for your help!

**Reply by sankarmaheshr** [reply]
Hi @cdiener .
I was using the above approach to simulate my communities. I observed that whenever I try to enforce an objective function that has 2 reactions (1 from each member of the community) then it throws an error 'NoneType' object has no attribute.
But when I simulate the community with just 1 reaction from one member of the community then it works without throwing any error.
What could possible be the reason for this?
![image](https://github.com/micom-dev/micom/assets/124859020/56c380a9-32d1-49f9-b6a5-a13c640bbc9c)
![image](https://github.com/micom-dev/micom/assets/124859020/b1d6a065-3110-4451-b061-6c113e2d7b48)
